### PR TITLE
change rabc code struct

### PR DESCRIPTION
**Pull request by yxxhero**
Signed-off-by: yxxhero <aiopsclub@163.com>

## Related Issue
https://github.com/dragonflyoss/Dragonfly2/issues/546

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [x] Code compiles correctly.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: manager/handlers/permission.go [hunk 1]
```diff
@@ -23,97 +23,110 @@ import (
...
+// @Description Get Role
 // @Tags permission
-// @Accept json
+// @Accept text
 // @Produce json
 // @Success 200
 // @Failure 400 {object} HTTPError
 // @Failure 500 {object} HTTPError
-// @Router /permission [post]
+// @Router /roles/:role_name [get]
```
Comment (gaius-qi) [inline manager/handlers/permission.go:121]: `roles/:id`

File: manager/handlers/permission.go [hunk 2]
```diff
@@ -23,97 +23,110 @@ import (
...
+// @Description Update Role by json config
+// @Tags role
+// @Accept json
 // @Produce json
-// @Success 200 {object}
+// @Success 200
 // @Failure 400 {object} HTTPError
 // @Failure 500 {object} HTTPError
-// @Router /permission/{subject}/{object}/{action} [get]
+// @Router /roles [post]
```
Comment (gaius-qi) [inline manager/handlers/permission.go:76]: `patch`

File: manager/handlers/permission.go [hunk 3]
```diff
@@ -23,97 +23,110 @@ import (
...
 // @Failure 400 {object} HTTPError
 // @Failure 500 {object} HTTPError
-// @Router /permission/roles/{subject} [get]
+// @Router /roles [post]
 
-func (h *Handlers) GetRolesForUser(ctx *gin.Context) {
-	var params types.UserRolesParams
-	if err := ctx.ShouldBindUri(&params); err != nil {
+func (h *Handlers) CreateRole(ctx *gin.Context) {
+	var json types.RolePermissionCreateRequest
```
Comment (gaius-qi) [inline manager/handlers/permission.go:55]: `CreatePermissionRequest`

File: manager/handlers/permission.go [hunk 4]
```diff
@@ -23,97 +23,110 @@ import (
...
-func (h *Handlers) HasRoleForUser(ctx *gin.Context) {
-	var params types.UserHasRoleParams
+	var params types.RoleParams
 	if err := ctx.ShouldBindUri(&params); err != nil {
 		ctx.JSON(http.StatusUnprocessableEntity, gin.H{"errors": err.Error()})
 		return
 	}
-	if params.Subject == "admin" {
-		ctx.JSON(http.StatusOK, gin.H{"has": true})
+	var json types.RolePermissionUpdateRequest
```
Comment (gaius-qi) [inline manager/handlers/permission.go:85]: `UpdateRoleRequest`

File: manager/handlers/permission.go [hunk 5]
```diff
@@ -23,97 +23,110 @@ import (
...
+// @Tags role
+// @Accept text
+// @Produce json
+// @Success 200
+// @Failure 400 {object} HTTPError
+// @Failure 500 {object} HTTPError
+// @Router /roles [get]
+
+func (h *Handlers) GetRoles(ctx *gin.Context) {
+	roles := h.Service.GetRoles()
```
Comment (gaius-qi) [inline manager/handlers/permission.go:109]: Add `Pagination`

File: manager/handlers/permission.go [hunk 6]
```diff
@@ -124,15 +137,15 @@ func (h *Handlers) CreatePermission(ctx *gin.Context) {
 // @Success 200
 // @Failure 400 {object} HTTPError
 // @Failure 500 {object} HTTPError
-// @Router /permission [delete]
+// @Router /role/:role_name [post]
```
Comment (gaius-qi) [inline manager/handlers/permission.go:140]: `:id` & `delete`

File: manager/handlers/user.go [hunk 7]
```diff
@@ -48,3 +48,75 @@ func (h *Handlers) SignUp(ctx *gin.Context) {
...
+// @Description Grant Role by uri config
+// @Tags users
+// @Accept json
+// @Produce json
+// @Success 200
+// @Failure 400 {object} HTTPError
+// @Failure 500 {object} HTTPError
+// @Router /users/:user_name/roles/:role_name [post]
+
+func (h *Handlers) GrantRole(ctx *gin.Context) {
```
Comment (gaius-qi) [inline manager/handlers/user.go:86]: `AddRoleToUser`

File: manager/handlers/user.go [hunk 8]
```diff
@@ -48,3 +48,75 @@ func (h *Handlers) SignUp(ctx *gin.Context) {
...
+
+// @Summary Grant Role
+// @Description Grant Role by uri config
+// @Tags users
+// @Accept json
+// @Produce json
+// @Success 200
+// @Failure 400 {object} HTTPError
+// @Failure 500 {object} HTTPError
+// @Router /users/:user_name/roles/:role_name [post]
```
Comment (gaius-qi) [inline manager/handlers/user.go:84]: `users/:id/roles/:role_id`

File: manager/handlers/user.go [hunk 9]
```diff
@@ -48,3 +48,75 @@ func (h *Handlers) SignUp(ctx *gin.Context) {
...
+}
+
+// @Summary Get User Roles
+// @Description Get User Roles
+// @Tags User
+// @Produce json
+// @Success 200 {object} RoutesInfo
+// @Failure 400 {object} HTTPError
+// @Failure 500 {object} HTTPError
+// @Router /user/:user_name/roles [get]
```
Comment (gaius-qi) [inline manager/handlers/user.go:107]: `users/:id/roles`

File: manager/handlers/user.go [hunk 10]
```diff
@@ -48,3 +48,75 @@ func (h *Handlers) SignUp(ctx *gin.Context) {
...
+// @Failure 500 {object} HTTPError
+// @Router /users/:user_name/roles/:role_name [post]
+
+func (h *Handlers) GrantRole(ctx *gin.Context) {
+	var params types.RoleRequest
+	if err := ctx.ShouldBindUri(&params); err != nil {
+		ctx.JSON(http.StatusUnprocessableEntity, gin.H{"errors": err.Error()})
+		return
+	}
+	err := h.Service.GrantRole(params.UserNmae, params.RoleName)
```
Comment (gaius-qi) [inline manager/handlers/user.go:92]: `AddRoleToUser`

File: manager/handlers/user.go [hunk 11]
```diff
@@ -48,3 +48,75 @@ func (h *Handlers) SignUp(ctx *gin.Context) {
...
+// @Failure 500 {object} HTTPError
+// @Router /user/:user_name/roles [get]
+
+func (h *Handlers) GetRolesForUser(ctx *gin.Context) {
+	var params types.UserParams
+	if err := ctx.ShouldBindUri(&params); err != nil {
+		ctx.JSON(http.StatusUnprocessableEntity, gin.H{"errors": err.Error()})
+		return
+	}
+	roles, err := h.Service.GetRolesForUser(params.UserName, ctx.GetString("userName"))
```
Comment (gaius-qi) [inline manager/handlers/user.go:115]: Add `Pagination`

File: manager/handlers/user.go [hunk 12]
```diff
@@ -48,3 +48,75 @@ func (h *Handlers) SignUp(ctx *gin.Context) {
...
+	if err != nil {
+		ctx.Error(err)
+		return
+	}
+	ctx.JSON(http.StatusOK, gin.H{"roles": roles})
```
Comment (gaius-qi) [inline manager/handlers/user.go:120]: ctx.JSON(http.StatusOK, roles)

File: manager/router/router.go [hunk 13]
```diff
@@ -58,7 +58,10 @@ func Init(verbose bool, service service.REST, enforcer *casbin.Enforcer) (*gin.E
 	ai.POST("/signin", jwt.LoginHandler)
 	ai.POST("/signout", jwt.LogoutHandler)
 	ai.POST("/refresh_token", jwt.RefreshHandler)
-	ai.POST("/signup", jwt.MiddlewareFunc(), rbac, h.SignUp)
+	ai.POST("/:user_name/roles/:role_name", h.GrantRole)
```
Comment (gaius-qi) [inline manager/router/router.go:61]: /:id/roles/:role_id

File: manager/router/router.go [hunk 14]
```diff
@@ -58,7 +58,10 @@ func Init(verbose bool, service service.REST, enforcer *casbin.Enforcer) (*gin.E
...
+	ai.DELETE("/:user_name/roles/:role_name", h.RevokeRole)
```
Comment (gaius-qi) [inline manager/router/router.go:62]: /:id/roles/:role_id

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: manager/handlers/user.go (hunk 11 above)
Comment (yxxhero) [reply manager/handlers/user.go]: The result return by casbin, it dou't support pagination.

File: manager/handlers/permission.go (hunk 5 above)
Comment (yxxhero) [reply manager/handlers/permission.go:109]: Controlled by casbin

**Review by gaius-qi — APPROVED** [review]
LGTM

**Review by 244372610 — APPROVED** [review]

**Pull request merged**